fixed the poolSubpage.maxNumElems == 1, free subpage bug

ZzxyNn opened the pull request:
Motivation:

when customer need large of 'byteBuf.capacity' in [7168, 8192], the size of 'chunk.subpages' may be inflated when large of byteBuf be released, not consistent with other 'byteBuf.capacity'

Modification:

when maxNumElems == 1 need consider remove from pool

Result:

Fixes #10896 .

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:
```diff
@@ -1,10 +1,31 @@
+
 import static org.junit.Assert.assertEquals;
 
 public class PoolSubpageReleaseTest {
```
unit tests should use junit

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:
```diff
@@ -1,10 +1,31 @@
+/*
+ * Copyright 2012 The Netty Project
```
2020

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:
```diff
@@ -0,0 +1,99 @@
+        try {
+            chunk = ReflectionUtils.tryToReadFieldValue(PooledByteBuf.class.getDeclaredField("chunk"), one).get();
+            Field subpagesField = PoolChunk.class.getDeclaredField("subpages");
+            return (PoolSubpage[]) ReflectionUtils.tryToReadFieldValue(subpagesField, chunk).get();
```
Please expose field as package private so it can be accessed here without reflection

ZzxyNn replied:
which field expose package private,  subpages ?

normanmaurer replied:
Yes

ZzxyNn replied:
for a unit test?

normanmaurer replied:
yes... Reflection should only be used as last resort... just make it package-private and add a comment that it is only package-private for testing.

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:
```diff
@@ -0,0 +1,99 @@
+        // release the 'ones'
+        for (int i = 0; i < oneLength; i++) {
+            ones[i].release();
+            // System.out.println("subpages size: " + countSubpages(subpages));
```
Remove

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:
```diff
@@ -0,0 +1,99 @@
+
+        //allocate
+        for (int i = 0; i < fixLength; i++) {
+            fixByteBuf[i] = ByteBufAllocator.DEFAULT.buffer(elemSize, elemSize);
```
You should directly instance a PooledByteBufAllocator here... or even merge this into PooledByteBufAllocatorTest

ZzxyNn commented:
@normanmaurer 
what’s else problem?

hyperxpro commented:
```
Run ./.github/scripts/check_leak.sh build-leak.output
Leak detected, please inspect build log
Error: Process completed with exit code 1.
```

Fix `ByteBuf` leaks.

ZzxyNn commented:
> ```
> Run ./.github/scripts/check_leak.sh build-leak.output
> Leak detected, please inspect build log
> Error: Process completed with exit code 1.
> ```
> 
> Fix `ByteBuf` leaks.

i tried to remove the unit tests i added, then the Leak detected passes, but in the unit tests i added , all ByteBuf calls release.

mostroverkhov commented:
@ZzxyNn 
There is `ClassCastException` in test in commit [675c4b](https://github.com/netty/netty/runs/1621650494), It is thrown before buffers release statements are reached
```
2020-12-29T12:25:52.3497652Z [ERROR] testNormalPoolSubpageRelease  Time elapsed: 0 s  <<< ERROR!
2020-12-29T12:25:52.3499472Z java.lang.ClassCastException: io.netty.buffer.AdvancedLeakAwareByteBuf cannot be cast to io.netty.buffer.PooledUnsafeHeapByteBuf
2020-12-29T12:25:52.3501741Z 	at io.netty.buffer.PooledByteBufAllocatorTest.run(PooledByteBufAllocatorTest.java:689)
2020-12-29T12:25:52.3504287Z 	at io.netty.buffer.PooledByteBufAllocatorTest.testNormalPoolSubpageRelease(PooledByteBufAllocatorTest.java:676)
```

ZzxyNn commented:
ok，its all checks passed

normanmaurer commented:
@ZzxyNn so I am back at work and had another look.

WDYT about applying this change on top of yours:

```
diff --git a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
index b3879ff433..4d7a5ebb9e 100644
--- a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
+++ b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
@@ -632,8 +632,8 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
     }
 
     @SuppressWarnings("unchecked")
-    private static PooledByteBuf<ByteBuffer> unwrapIfNeeded(ByteBuf buf) {
-        return (PooledByteBuf<ByteBuffer>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
+    private static <T> PooledByteBuf<T> unwrapIfNeeded(ByteBuf buf) {
+        return (PooledByteBuf<T>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
     }
 
     @Test
@@ -667,27 +667,17 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
         // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
         // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
         int elemSize = 32;
-        run(elemSize, false);
-    }
-
-    private void run(int elemSize, boolean preferDirect) {
-        int oneLength = 16, fixedLength = 256;
+        int oneLength = 16;
+        int fixedLength = 256;
         ByteBuf[] ones = new ByteBuf[oneLength];
         ByteBuf[] fixedByteBuf = new ByteBuf[fixedLength];
-        PooledByteBufAllocator allocator = newAllocator(preferDirect);
+        final PooledByteBufAllocator allocator = newAllocator(false);
 
-        //allocate
         for (int i = 0; i < fixedLength; i++) {
             fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
         }
 
-        if (!PooledByteBuf.class.isAssignableFrom(fixedByteBuf[0].getClass())) {
-            for (int i = 0; i < fixedLength; i++) {
-                fixedByteBuf[i].release();
-            }
-            return;
-        }
-        PoolSubpage[] subpages = ((PooledByteBuf) fixedByteBuf[0]).chunk.subpages;
+        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
 
         // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
         // the chunk.poolSubpages.size
@@ -697,7 +687,7 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
             ones[i] = allocator.heapBuffer(elemSize, elemSize);
         }
 
-        //first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
+        // first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
         // ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
         for (int i = 0; i < fixedLength; i++) {
             fixedByteBuf[i].release();
@@ -711,10 +701,10 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
         assertEquals(countSubpages(subpages), fixedSubpagesCount + 1);
     }
 
-    private static int countSubpages(PoolSubpage<Object>[] subpages) {
+    private static int countSubpages(PoolSubpage<?>[] subpages) {
         int num = 0;
-        for (final PoolSubpage<Object> subpage : subpages) {
-            if (null != subpage) {
+        for (final PoolSubpage<?> subpage : subpages) {
+            if (subpage != null) {
                 num++;
             }
         }

```

ZzxyNn commented:
> @ZzxyNn so I am back at work and had another look.
> 
> WDYT about applying this change on top of yours:
> 
> ```
> diff --git a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> index b3879ff433..4d7a5ebb9e 100644
> --- a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> +++ b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> @@ -632,8 +632,8 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>      }
>  
>      @SuppressWarnings("unchecked")
> -    private static PooledByteBuf<ByteBuffer> unwrapIfNeeded(ByteBuf buf) {
> -        return (PooledByteBuf<ByteBuffer>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
> +    private static <T> PooledByteBuf<T> unwrapIfNeeded(ByteBuf buf) {
> +        return (PooledByteBuf<T>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
>      }
>  
>      @Test
> @@ -667,27 +667,17 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>          // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
>          // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
>          int elemSize = 32;
> -        run(elemSize, false);
> -    }
> -
> -    private void run(int elemSize, boolean preferDirect) {
> -        int oneLength = 16, fixedLength = 256;
> +        int oneLength = 16;
> +        int fixedLength = 256;
>          ByteBuf[] ones = new ByteBuf[oneLength];
>          ByteBuf[] fixedByteBuf = new ByteBuf[fixedLength];
> -        PooledByteBufAllocator allocator = newAllocator(preferDirect);
> +        final PooledByteBufAllocator allocator = newAllocator(false);
>  
> -        //allocate
>          for (int i = 0; i < fixedLength; i++) {
>              fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
>          }
>  
> -        if (!PooledByteBuf.class.isAssignableFrom(fixedByteBuf[0].getClass())) {
> -            for (int i = 0; i < fixedLength; i++) {
> -                fixedByteBuf[i].release();
> -            }
> -            return;
> -        }
> -        PoolSubpage[] subpages = ((PooledByteBuf) fixedByteBuf[0]).chunk.subpages;
> +        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
>  
>          // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
>          // the chunk.poolSubpages.size
> @@ -697,7 +687,7 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>              ones[i] = allocator.heapBuffer(elemSize, elemSize);
>          }
>  
> -        //first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
> +        // first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
>          // ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
>          for (int i = 0; i < fixedLength; i++) {
>              fixedByteBuf[i].release();
> @@ -711,10 +701,10 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>          assertEquals(countSubpages(subpages), fixedSubpagesCount + 1);
>      }
>  
> -    private static int countSubpages(PoolSubpage<Object>[] subpages) {
> +    private static int countSubpages(PoolSubpage<?>[] subpages) {
>          int num = 0;
> -        for (final PoolSubpage<Object> subpage : subpages) {
> -            if (null != subpage) {
> +        for (final PoolSubpage<?> subpage : subpages) {
> +            if (subpage != null) {
>                  num++;
>              }
>          }
> ```

emmm, clearly and simply

normanmaurer commented:
Please apply then

ZzxyNn commented:
> Please apply then

it's done

chrisvest commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -661,4 +661,53 @@ public void testCacheWorksForNormalAllocations() {
+            ones[i] = allocator.heapBuffer(elemSize, elemSize);
+        }
+
+        // first release the 'fixedByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
```
```suggestion
        // first release the 'fixedByteBuf', in order to fill the PoolThreadCache.MemoryRegionCache.queue
```

ZzxyNn replied:
Has been adjusted.

chrisvest commented on buffer/src/main/java/io/netty/buffer/PoolSubpage.java:
```diff
@@ -115,7 +115,9 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
-            return true;
+            if (maxNumElems > 1) {
+                return true;
+            }
```
The test you added passes without this change. So this is not really being tested.

ZzxyNn replied:
Has been adjusted.

chrisvest commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -661,4 +661,53 @@ public void testCacheWorksForNormalAllocations() {
+            fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
+        }
+
+        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
```
Try to write a test that verifies on the externally observable behaviour of the allocator, rather than asserting on internal state changes.

Tests that rely on implementation internals tend to be more brittle, and easier to break on unrelated changes.

ZzxyNn replied:
the externally observable behavior of the allocator is when 'maxNumElems == 1' then release subpage operate not work, have no idea how to asserting that @chrisvest

chrisvest reviewed (changes requested):
The test seems to be inadequate.

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
+        for (int i = 0; i < length; i++) {
+            byteBufs[i].release();
         }
+        int AfterFreeBytes = chunk.freeBytes();
```
nit: s/After.../after../

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
-            ones[i] = allocator.heapBuffer(elemSize, elemSize);
+        PooledByteBuf<Object> pooledByteBuf = unwrapIfNeeded(byteBufs[0]);
+        // disable PoolThreadCache
+        pooledByteBuf.cache = null;
```
don't do this... you can construct the `PooledByteBufAllocator` with caches disabled.

ZzxyNn replied:
OK, please recheck.

ZzxyNn replied:
@normanmaurer How's it going?

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
+            byteBufs[i] = allocator.heapBuffer(elemSize, elemSize);
         }
-
-        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
```
I think this means that `subpages` can be private again ?

chrisvest commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -661,4 +665,27 @@ public void testCacheWorksForNormalAllocations() {
+    public void testNormalPoolSubpageRelease() {
+        // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
+        // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
+        int elemSize = 1024;
```
If you make `elemSize = 8192` then this test will fail when your `PoolSubpage` changes are commented out.

```suggestion
        int elemSize = 8192;
```

ZzxyNn replied:
yes, is there a problem? @chrisvest

ZzxyNn replied:
If my change is commented and the unit test fails, doesn't that mean the change is correct

chrisvest replied:
Yes. My point is: when your `PoolSubpage` changes are commented out, the test does _not_ fail when the `elemSize` is 1024, but it _does_ fail when `elemSize` is 8192, which is why I'm suggesting making this change.

With your `PoolSubpage` changes in place, the test of course pass for either `elemSize`.

ZzxyNn replied:
oh, I see what you mean. I did this before, but you said that my changes were not covered by the unit tests

chrisvest reviewed (changes requested):
Just one thing

ZzxyNn commented:
@chrisvest 
`Error: Cannot retrieve repository metadata (repomd.xml) for repository: base. Please verify its path and try again.`

What's going on here

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolSubpage.java:
```diff
@@ -115,7 +115,9 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
         if (numAvail ++ == 0) {
             addToPool(head);
-            return true;
+            if (maxNumElems > 1) {
```
@ZzxyNn can you just please add a comment here any explain why we return true only if `maxNumElems > 1` ? This will make it easier to understand this in the future.

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:
```diff
@@ -197,6 +198,13 @@ public PooledByteBufAllocator(boolean preferDirect) {
+                DEFAULT_SMALL_CACHE_SIZE, DEFAULT_NORMAL_CACHE_SIZE, useCacheForAllThreads,
+                DEFAULT_DIRECT_MEMORY_CACHE_ALIGNMENT);
+    }
+
```
Basically all the changes in this class should be reverted

normanmaurer commented on buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:
```diff
@@ -661,4 +665,27 @@ public void testCacheWorksForNormalAllocations() {
+        int elemSize = 8192;
+        int length = 1024;
+        ByteBuf[] byteBufs = new ByteBuf[length];
+        final PooledByteBufAllocator allocator = newAllocator(false, false);
```
The above

ZzxyNn replied:
please review..

normanmaurer commented on buffer/src/main/java/io/netty/buffer/PoolSubpage.java:
```diff
@@ -115,7 +115,10 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
         if (numAvail ++ == 0) {
             addToPool(head);
-            return true;
+            // when maxNumElems == 1, cannot return directly, also need to go through the following judgment
```
Please reword... This comment doesn't really explain what this is all about.

ZzxyNn replied:
If `maxNumeLems == 1`, the maximum `numAvail` is also 1, and each free operation of the `PoolSubpage` will return in the branch of `numAvail ++ == 0`, which means the following code is unreachable

how about this?

ZzxyNn replied:
If `maxNumElems == 1`, the maximum `numAvail` is also 1. The free operation of this type of `PoolSubPage` will end here, which means that subsequent code is not reachable, these `PoolSubPages` will not really be recycled, so the `maxNumElems > 1` judgment needs to be added

normanmaurer replied:
sounds good... Please update the comment to something like this^^^

ZzxyNn replied:
alright.

ZzxyNn commented:
`transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollSocketShutdownOutputBySelfTest.java:1 | expected null, but was:<true>`

What's going on here?

normanmaurer commented:
@ZzxyNn flaky test... triggered the CI again

ZzxyNn commented:
> @ZzxyNn flaky test... triggered the CI again

`Error: Resource not accessible by integration`
em....?

normanmaurer commented:
This will be fixed by https://github.com/netty/netty/pull/10968... I will re-trigger the tests when this is pulled in.

ZzxyNn commented:
> This will be fixed by #10968... I will re-trigger the tests when this is pulled in.

fine.

normanmaurer commented:
@ZzxyNn can you rebase so the CI changes are pulled in as well. ?

normanmaurer commented:
@chrisvest can you have a final look ?

normanmaurer commented:
@ZzxyNn Ok I think this one is ready to go now... Can you please squash the commits and adjust the commit message to explain exactly what this fixes etc ?

ZzxyNn commented:
> @ZzxyNn Ok I think this one is ready to go now... Can you please squash the commits and adjust the commit message to explain exactly what this fixes etc ?

Oh...I feel like I'm having a problem with my merge.

ZzxyNn commented:
@normanmaurer em......?

normanmaurer commented:
@ZzxyNn seems like the commits were not squashed yet and the commit message adjusted... Can you please do this ?

normanmaurer commented:
See https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

ZzxyNn commented:
> @ZzxyNn seems like the commits were not squashed yet and the commit message adjusted... Can you please do this ?

little hard for me...

ZzxyNn commented:
holy ...

chrisvest commented:
@ZzxyNn Yeah, rebasing can be troublesome when there are merge commits involved. If it's easier for you to make a new branch with the final changes, then just do that.
